### PR TITLE
[spirv] Buffer/RWbuffer Load() and operator[]

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

**Comment by ehsannas** [comment]
@antiagainst , please review

**Comment by ehsannas** [comment]
Touches only SPIR-V code

**Comment by antiagainst** [comment]
Just wanted to note that from this pull request and on, we will do our reviews here directly. Please wait until either @ehsannas or @antiagainst gives an approval before merging.

And it seems we cannot add reviewers by ourselves. So please feel free to add us as the reviewers for all future SPIR-V related pull requests. :)

File: tools/clang/unittests/SPIRV/CodeGenSPIRVTest.cpp [hunk 1]
```diff
@@ -343,6 +343,12 @@ TEST_F(FileTest, ByteAddressBufferStore) {
   runFileTest("method.byte-address-buffer.store.hlsl");
 }
 
+// For Buffer/RWBuffer methods
+TEST_F(FileTest, BufferLoad) { runFileTest("buffer.load.hlsl"); }
+TEST_F(FileTest, BufferOperatorBrackets) {
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/CodeGenSPIRVTest.cpp]: This is named as "access" previously and put around Line 200. I'd suggest to rename it as `op.buffer.access.hlsl` to keep consistent with others.

File: tools/clang/lib/SPIRV/TypeTranslator.h [hunk 2]
```diff
@@ -52,6 +52,12 @@ class TypeTranslator {
   /// \brief Returns true if the given type is the HLSL RWByteAddressBufferType.
   bool isRWByteAddressBuffer(QualType type);
 
+  /// \brief Returns true if the given type is the HLSL Buffer type.
+  bool isBuffer(QualType type);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.h]: Can be static.

File: tools/clang/lib/SPIRV/TypeTranslator.h [hunk 3]
```diff
@@ -52,6 +52,12 @@ class TypeTranslator {
...
+
+  /// \brief Returns true if the given type is the HLSL RWBuffer type.
+  bool isRWBuffer(QualType type);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.h]: Can be static.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 4]
```diff
@@ -1111,7 +1111,8 @@ uint32_t SPIRVEmitter::doCastExpr(const CastExpr *expr) {
   switch (expr->getCastKind()) {
   case CastKind::CK_LValueToRValue: {
     const uint32_t fromValue = doExpr(subExpr);
-    if (isVectorShuffle(subExpr) || isa<ExtMatrixElementExpr>(subExpr)) {
+    if (isVectorShuffle(subExpr) || isa<ExtMatrixElementExpr>(subExpr) ||
+        isBufferIndexing(dyn_cast<CXXOperatorCallExpr>(subExpr))) {
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/TypeTranslator.h (hunk 2 above)
Comment (ehsannas) [inline tools/clang/lib/SPIRV/TypeTranslator.h]: Done

File: tools/clang/lib/SPIRV/TypeTranslator.h (hunk 3 above)
Comment (ehsannas) [inline tools/clang/lib/SPIRV/TypeTranslator.h]: Done

File: tools/clang/unittests/SPIRV/CodeGenSPIRVTest.cpp (hunk 1 above)
Comment (ehsannas) [inline tools/clang/unittests/SPIRV/CodeGenSPIRVTest.cpp]: Done

**Pull request merged**